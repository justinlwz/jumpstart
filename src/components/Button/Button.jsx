import React, { Component } from 'react'
import PropTypes from 'prop-types'

import ButtonStyle from './Button.style'

class Button extends Component {

    // PropTypes
    static propTypes = {
        children: PropTypes.node,
        className: PropTypes.string,
        disabled: PropTypes.bool,
        href: PropTypes.string,
        onClick: PropTypes.func,
        small: PropTypes.bool,
        target: PropTypes.oneOf([
            '_blank',
            '_parent',
            '_self',
            '_top',
        ]),
    }

    _handleClick = (e) => {
        const { disabled, onClick } = this.props
        if (disabled) {
            e.preventDefault()
            return
        }
        if (onClick) onClick(e)
    }

    render() {

        // Props
        const {
            children,
            href,
            ...filteredProps
        } = this.props

        // Create Element
        if (href) {
            return(
                <ButtonStyle 
                    as="a"
                    href={href} 
                    rel="noreferrer noopener"
                    {...filteredProps}
                >
                    {children}
                </ButtonStyle>
            )
        }
        else {
            return (
                <ButtonStyle as="button" {...filteredProps}>
                    {children}
                </ButtonStyle>
            )
        }
    }
}

export default Button
