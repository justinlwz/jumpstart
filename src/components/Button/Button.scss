/*


    Buttons


*/



/* -------------------------------------

    Imports

------------------------------------- */

@import "../Styles/core.scss";


/* -------------------------------------

    Variables

------------------------------------- */

// Animation Timing
$btn-anim-timing: 200ms;

// Large
$btn-height:      40px !default;
$btn-icon-size:   40px !default;
$btn-font-size:   12px !default;

// Small
$btn-small-height:      32px !default;
$btn-icon-small-size:   32px !default;
$btn-small-font-size:   10px !default;


/* -------------------------------------

    Mixins

------------------------------------- */

@mixin button-setter(
    $width: auto,
    $height: $btn-height,
    $side-padding: 16px,
    $mode: normal
) {
    // Size
    width: $width;
    height: $height;
    // Appearance
    position: relative;
    display: inline-block;
    vertical-align: middle;

    // Padding
    @if $mode == normal {
        text-align: center;
        padding: 0 $side-padding;
    }
    @else if $mode == left {
        text-align: left;
        padding: 0 $side-padding 0 $height;
    }
    @else if $mode == right {
        text-align: right;
        padding: 0 $height 0 $side-padding;
    }
    @else if $mode == icon {
        text-align: center;
        padding: 0;
    }
}


/* -------------------------------------

    Button Styles

------------------------------------- */

// Base Class
.btn {
    font-weight: 600;
    line-height: 1;
    position: relative;
    cursor: pointer;
    outline: none;
    border: none;
    background-color: transparent;
    text-decoration: none;
    text-transform: uppercase;
    &:disabled { pointer-events: none; }
}

.btn.solid {
    color: white;
    background-color: $color-nav;
    @include transition($prop: background-color, $timing: $btn-anim-timing);
    &:hover {
        background-color: $color-nav-hover;
    }
    &:disabled {
        background-color: $color-gray-D00;
    }
}

.btn.line {
    color: $color-nav;
    outline: solid 1px $color-gray-D00;
    @include transition($prop: color, $timing: $btn-anim-timing);
    &:hover {
        color: $color-nav-hover;
        outline-width: 2px;
    }
    &:disabled {
        color: $color-gray-D00;
        outline-color: $color-gray-D00;
    }
    &.icon-text {
        color: $color-gray-400;
        i { color: $color-nav; }
        &:hover i { color: $color-nav-hover; }
    }
}

.btn.quiet {
    color: $color-gray-400;
    transition: color $btn-anim-timing ease-in-out,
                background-color $btn-anim-timing ease-in-out;
    &:hover {
        color: $color-nav-hover;
        background-color: rgba($color-nav-hover, 0.10);
    }
    &:disabled {
        color: $color-gray-D00;
    }
    &.icon-text {
        color: $color-gray-400;
        i { color: $color-nav; }
        &:hover i { color: $color-nav-hover; }
    }
}

.btn.no-effect {
    color: $color-gray-400;
}


/* -------------------------------------

    Button Types

------------------------------------- */

.btn {
    font-size: $btn-font-size;
    @include button-setter($mode: normal);
}

.btn.link {
    line-height: $btn-height;
    &.line {
        line-height: $btn-height;
        &:hover { line-height: $btn-height; }
    }
    &.small {
        line-height: $btn-small-height;
        &.line {
            line-height: $btn-small-height;
            &:hover { line-height: $btn-small-height; }
        }
    }
    &.icon i,
    &.icon img {
        top: 50%;
        left: 50%;
        transform: translate(-50%, -50%);
        position: absolute;
    }
}

.btn.small {
    font-size: $btn-small-font-size;
    @include button-setter($height: $btn-small-height, $mode: normal);
}

.btn.icon {
    @include button-setter($width: 64px, $height: 64px, $mode: icon);
    img {
       width: auto;
       height: 20px;
       display: inline-block;
    }
    i, img {
        vertical-align: middle;
    }
}

.btn.circle {
    @include round-corners($value: 50%);
}

.btn.icon-text {
    &.icon-left {
        @include button-setter($mode: left);
        i {
            top: 0;
            left: 0;
        }
    }
    &.icon-right {
        @include button-setter($mode: right);
        i {
            top: 0;
            right: 0;
        }
    }
    &.line i {
        height: $btn-icon-size;
        line-height: $btn-icon-size;
    }
    &.line:hover i {
        height: $btn-icon-size;
        line-height: $btn-icon-size;
    }
    i {
        width: $btn-icon-size;
        height: $btn-icon-size;
        line-height: $btn-icon-size;
        text-align: center;
        display: block;
        position: absolute;
    }
    span {
        vertical-align: middle;
        display: inline-block;
    }
}

.btn.icon-text.small {
    &.icon-left {
        @include button-setter($height: $btn-small-height, $mode: left);
        i {
            top: 0;
            left: 0;
        }
    }
    &.icon-right {
        @include button-setter($height: $btn-small-height, $mode: right);
        i {
            top: 0;
            right: 0;
        }
    }
    &.line i {
        height: $btn-icon-small-size;
        line-height: $btn-icon-small-size;
    }
    &.line:hover i {
        height: $btn-icon-small-size;
        line-height: $btn-icon-small-size;
    }
    i {
        width: $btn-icon-small-size;
        height: $btn-icon-small-size;
        font-size: 16px;
        line-height: $btn-icon-small-size;
        text-align: center;
        display: block;
        position: absolute;
    }
    span {
        vertical-align: middle;
        display: inline-block;
    }
}
