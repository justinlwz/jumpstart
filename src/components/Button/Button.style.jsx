import styled, { css } from 'styled-components'
import { transition } from '../../components/Styles/StyleUtils/Mixins.style'
import { nav, navActive, disabled } from '../../components/Styles/StyleUtils/Colours.style'

const ButtonStyle = styled.button`
    height: ${props => props.small ? "24px" : "32px"};
    font-size: ${props => props.small ? "12px" : "14px"};
    font-weight: 600;
    line-height: ${props => props.small ? "24px" : "32px"};
    position: relative;
    cursor: pointer;
    padding: ${props => props.small ? "0 8px" : "0 16px"};
    border-radius: ${props => props.radius + 'px'};
    outline: none;
    border: none;
    text-decoration: none;
    ${transition({timing:'150ms'})}

    /* Disabled Style */
    :disabled {  pointer-events: none; }

    /* Button Type */
    ${props => props.type == 'normal' && css`
        color: white;
        background-color: ${nav};
        :hover {
            background-color: ${navActive};
        }
        :disabled {
            background-color: ${disabled};
        }
    `}

    ${props => props.type == 'quiet' && css`
        color: ${nav};
        background-color: transparent;
        :hover {
            color: ${navActive};
            background-color: rgba(19, 161, 228, 0.05);
        }
        :disabled {
            color: ${disabled};
        }
    `}

`

export default ButtonStyle