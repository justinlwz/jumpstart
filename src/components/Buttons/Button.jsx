import React, { Component } from 'react';
import PropTypes from 'prop-types';


class Button extends Component {

    // PropTypes
    static propTypes = {
        children: PropTypes.node,
        className: PropTypes.string,
        disabled: PropTypes.bool,
        href: PropTypes.string,
        onClick: PropTypes.func,
        small: PropTypes.bool,
        target: PropTypes.oneOf([
            '_blank',
            '_parent',
            '_self',
            '_top',
        ]),
    };

    _handleClick = (e) => {
        const { disabled, onClick } = this.props;
        if (disabled) {
            e.preventDefault();
            return;
        }
        if (onClick) onClick(e);
    }

    render() {

        // Props
        const {
            children,
            href,
            ...filteredProps
        } = this.props;

        // Create Element
        if (href) {
            return(
                <a {...filteredProps} href={href} rel="noreferrer noopener">
                    {children}
                </a>
            );
        }
        else {
            return (
                <button {...filteredProps}>
                    {children}
                </button>
            );
        }
    }
};

export default Button;
