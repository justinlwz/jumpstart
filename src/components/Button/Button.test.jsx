import React from 'react';
import renderer from 'react-test-renderer';

import Button from './Button';

test('Button component mounts properly', () => {
    const wrapper = renderer.create(
        <Button onClick={ () => {alert('clicked!');} } type="solid">Click Me!</Button>
    );
    let tree = wrapper.toJSON();
    expect(tree).toMatchSnapshot();
});