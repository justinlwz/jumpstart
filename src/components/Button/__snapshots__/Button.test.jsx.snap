// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component mounts properly 1`] = `
<button
  className="sc-bdVaJa dbnRmD"
  onClick={[Function]}
  type="solid"
>
  Click Me!
</button>
`;
